File: src/ExtraUtilities/worlds/blocks/effect/Breaker.java
package ExtraUtilities.worlds.blocks.effect;

import arc.graphics.Color;
import arc.graphics.g2d.Font;
import arc.math.Mathf;
import arc.math.geom.Geometry;
import arc.math.geom.Rect;
import arc.util.Align;
import arc.util.Time;
import arc.util.Tmp;
import mindustry.Vars;
import mindustry.gen.Building;
import mindustry.graphics.Drawf;
import mindustry.graphics.Pal;
import mindustry.ui.Fonts;
import mindustry.world.Block;
import mindustry.world.Tile;

public class Breaker extends Block {
    public float timerBreak;
    public int maxsize;

    public Breaker(String name) {
        super(name);
        size = 1;
        rotate = true; //旋转找目标
        drawArrow = false;
        update = //设置了自爆，update是必要的，基本update设置true就行了
                solid = //设置固态，其实没必要
                        destructible = true;
        maxsize = 1;
        timerBreak = 300;
    }

    @Override
    public void drawPlace(int x, int y, int rotation, boolean valid) {
        super.drawPlace(x, y, rotation, valid);
        x *= Vars.tilesize;
        y *= Vars.tilesize;
        x += offset;
        y += offset;
        Rect rect = Tmp.r1;
        rect.setCentered(x, y, maxsize * Vars.tilesize);
        int len = Vars.tilesize * maxsize;

        rect.x += Geometry.d4x(rotation) * len;
        rect.y += Geometry.d4y(rotation) * len;

        Drawf.dashRect(valid ? Pal.accent : Pal.remove, rect);
    }

    public class BreakerBuild extends Building{
        public float Timer = 0;

        @Override
        public void updateTile() {
            Tile tile = Vars.world.tile(tileX() + Geometry.d4x(rotation), this.tileY() + Geometry.d4y(rotation));
            if(tile != null && tile.block() != null && tile.build == null && tile.block().solid && !tile.block().breakable && tile.block().size <= maxsize){
                Timer += Time.delta;
                if(Timer >= timerBreak){
                    Vars.world.tile(tile.x, tile.y).setAir();
                    kill();
                }
            }
        }

        @Override
        public void draw() {
            super.draw();
            int s = Mathf.ceil((timerBreak - Timer)/60);
            String  text = "|" + s + "|";

            Fonts.def.draw(text, this.x, this.y, Color.red, 0.35f, true, Align.center);
        }
    }
}

File: assets/scripts/block/effect/breaker.js
/** @author guiY */

const timerBreak = 180;//自爆时限 60 * 时间，单位：秒
const maxsize = 1;//默认就是一格，不建议改

const breaker = extend(Block, "breaker", {
    drawPlace(x, y, rotation, valid){
        this.super$drawPlace(x, y, rotation, valid);

        x *= Vars.tilesize;
        y *= Vars.tilesize;
        x += this.offset;
        y += this.offset;
        var rect = Tmp.r1;
        rect.setCentered(x, y, maxsize * Vars.tilesize);
        var len = Vars.tilesize * maxsize;

        rect.x += Geometry.d4x[rotation] * len;
        rect.y += Geometry.d4y[rotation] * len;

        Drawf.dashRect(valid ? Pal.accent : Pal.remove, rect);
    }
});
breaker.size = 1;
breaker.category = Category.effect;
breaker.requirements = ItemStack.with();
breaker.buildVisibility = BuildVisibility.shown;
breaker.rotate = true; //旋转找目标
breaker.drawArrow = false;
breaker.update = //设置了自爆，update是必要的，基本update设置true就行了
    breaker.solid = //设置固态，其实没必要
        breaker.destructible = true;
breaker.rebuildable = false;//设置不可重建，建议false
breaker.buildType = prov(() => {
    var timer = 0;

    return extend(Building, {
        updateTile() {
            this.bk();
        },

        bk(){
            var tile = Vars.world.tile(this.tileX() + Geometry.d4x[this.rotation], this.tileY() + Geometry.d4y[this.rotation]);
            if(tile != null && tile.block() != null && tile.build == null && tile.block().solid && !tile.block().breakable && tile.block().size <= maxsize){
                timer += Time.delta;
                if(timer >= timerBreak){
                    Call.removeTile(tile);
                    this.kill();
                }
            }
        },

        draw(){
            this.super$draw();
            var s = Math.ceil((timerBreak - timer)/60);
            var text = "|" + s + "|"

            Fonts.def.draw(text, this.x, this.y, Color.red, 0.35, true, Align.center);
        }
    });
})